stix2/markings/granular_markings.py

```diff
@@ -0,0 +1,269 @@
+
+    for marking in granular_markings:
+        for user_selector in selectors:
+            for marking_selector in marking.get("selectors", []):
+                if any([(user_selector == marking_selector),  # Catch explicit selectors.
+                        (user_selector.startswith(marking_selector) and inherited),  # Catch inherited selectors.
+                        (marking_selector.startswith(user_selector) and descendants)]):  # Catch descendants selectors
+                    refs = marking.get("marking_ref", [])
+                    results.update([refs])
+
+    return list(results)
+
+
+def set_markings(obj, selectors, marking):
+    """
+    Removes all markings associated with selectors and appends a new granular
+    marking. Refer to `clear_markings` and `add_markings` for details.
+
+    Args:
+        obj: An SDO or SRO object.
+        selectors: string or list of selector strings relative to the SDO or
+            SRO in which the properties appear.
+        marking: identifier or list of marking identifiers that apply to the
+            properties selected by `selectors`.
+
+    Returns:
+        A new version of the given SDO or SRO with specified markings removed
+        and new ones added.
+
+    """
+    obj = clear_markings(obj, selectors)
+    return add_markings(obj, selectors, marking)
+
+
+def remove_markings(obj, selectors, marking):
+    """
+    Removes granular_marking from the granular_markings collection.
+
+    Args:
+        obj: An SDO or SRO object.
+        selectors: string or list of selectors strings relative to the SDO or
+            SRO in which the properties appear.
+        marking: identifier or list of marking identifiers that apply to the
+            properties selected by `selectors`.
+
+    Raises:
+        InvalidSelectorError: If `selectors` fail validation.
+        MarkingNotFoundError: If markings to remove are not found on
+            the provided SDO or SRO.
+
+    Returns:
+        A new version of the given SDO or SRO with specified markings removed.
+
+    """
+    selectors = utils.convert_to_list(selectors)
+    utils.validate(obj, selectors)
+
+    granular_markings = obj.get("granular_markings")
+
+    if not granular_markings:
+        return obj
+
+    granular_markings = utils.expand_markings(granular_markings)
+
+    if isinstance(marking, list):
+        to_remove = []
+        for m in marking:
+            to_remove.append({"marking_ref": m, "selectors": selectors})
+    else:
+        to_remove = [{"marking_ref": marking, "selectors": selectors}]
+
+    remove = utils.build_granular_marking(to_remove).get("granular_markings")
+
+    if not any(marking in granular_markings for marking in remove):
+        raise exceptions.MarkingNotFoundError(obj, remove)
+
+    granular_markings = [
+        m for m in granular_markings if m not in remove
+    ]
+
+    granular_markings = utils.compress_markings(granular_markings)
+
+    if granular_markings:
```

Comment:
In order to make this work on dictionaries, we might need to reimplement `new_version` (or not, since in general our attitude is to trust people who build their own dictionaries to do the right thing).

Reply:
This has been fixed.